Memory Stream Refactor for PNGs

Leonardo-Ferreira opened the pull request:
Focusing on avoiding allocation of `MemoryStream` adhoc-ally I implemented the Microsoft recommend `RecyclableMemoryStream` which I described on #18721 but only for the PNG class thus far. The results of change alone are promising (GC collection reduction of about 10%) but the refactor I made right after made the memory footprint drop 2-5% and consequentially the GC collection dropped about 15%.

If this is approved I intend on spread the implementation to the whole code.

Leonardo-Ferreira committed:
New singleton class to handle MemoryStream distrib
Added package ref

Leonardo-Ferreira committed:
Implemented the MemoryStreamManager for PNGs

Leonardo-Ferreira committed:
Single Stream Read for length and type of png

Leonardo-Ferreira committed:
Refactored to reduce even further the allocation of MemoryStreams

Mailaender commented:
~~I tried to test this with @OpenHV which is a mod that solely uses .png sprites, but the benchmark mode crashes the game. https://github.com/OpenRA/OpenRA/issues/18725~~ 
I am actually not even sure if we benchmark anything that measures these particular code changes.

Leonardo-Ferreira committed:
Fine tuned the stream manager

Leonardo-Ferreira commented:
<img width="793" alt="evidence" src="https://user-images.githubusercontent.com/5585339/96151957-5ec07100-0ee2-11eb-8914-af355ab3de0d.png">
Here is some evidence of the improvements i mentioned above. I used the Visual Studio Enterprise "Diagnostic Tools" to extract the charts...
THe first one is the one with my proposed changes and the second one is the current "bleed" branch

Leonardo-Ferreira commented:
@Mailaender rebased!

Leonardo-Ferreira committed:
Whitelisted Microsoft.IO.RecyclableMemoryStream

RoosterDragon commented:
What did you do in the game to test the results of this change?

During loading, the `Png` constructor is called several times when the maps load. The calls to the `new MemoryStream(byte[] buffer)` overload uses the provided buffer as the backing array in the memory stream. Since the memory stream doesn't have to allocate its own buffer in the first place, I'm not aware of what benefit using `MemoryStreamManager` would provide here - could you explain?

The calls to `new MemoryStream()` in the `Save` method make sense as something that would benefit - but saving a PNG seems pretty rare - is there a scenario in the game you were trying to improve here?

pchote commented on Makefile:
```diff
@@ -39,7 +39,7 @@
 
 # These are explicitly shipped alongside our core files by the packaging script
-WHITELISTED_THIRDPARTY_ASSEMBLIES = ICSharpCode.SharpZipLib.dll FuzzyLogicLibrary.dll Eluant.dll BeaconLib.dll Open.Nat.dll SDL2-CS.dll OpenAL-CS.Core.dll DiscordRPC.dll Newtonsoft.Json.dll
+WHITELISTED_THIRDPARTY_ASSEMBLIES = ICSharpCode.SharpZipLib.dll FuzzyLogicLibrary.dll Eluant.dll BeaconLib.dll Open.Nat.dll SDL2-CS.dll OpenAL-CS.Core.dll DiscordRPC.dll Newtonsoft.Json.dll Microsoft.IO.RecyclableMemoryStream.dll
```
Changes to this list must be coordinated with changes to the `make install` target and packaging scripts to actually ship the assembly!

Leonardo-Ferreira replied:
Please advise on how to proceed

pchote replied:
See 9c6e7cc5d134bf891fb59863aa3e0bdcdcd16eb3 for an example of the required changes.

pchote commented on OpenRA.Game/MemoryStreamManager.cs:
```diff
@@ -0,0 +1,44 @@
+﻿using System;
```
All source files must start with OpenRA's standard license header.

Leonardo-Ferreira replied:
done!

pchote reviewed (commented):
A couple of extra points, and I second RoosterDragon's questions - we should be only using MemoryStreams behind the load screen, where we are already generating significant amounts of garbage, which is cleaned up with a forced GC before the gameplay phase starts. I would like to see some numbers quantifying the actual ingame benefit before we start any major efforts to rework the use of streams through the engine.

Leonardo-Ferreira commented:
> What did you do in the game to test the results of this change?

I ran the game with visual studio diagnostics plugged in. Memory execution profile posted above serves as evidence of the decrease activity on the garbage collector and the smaller memory footprint at start-up. The cleanest way to check the result is to actually dump the memory and perform a memory count but that would be VERY difficult because there are several memory streams all around and it be very hard to tell if that particular memory belong to a PNG or to another class...

> During loading, the `Png` constructor is called several times when the maps load. The calls to the `new MemoryStream(byte[] buffer)` overload uses the provided buffer as the backing array in the memory stream. Since the memory stream doesn't have to allocate its own buffer in the first place, I'm not aware of what benefit using `MemoryStreamManager` would provide here - could you explain?

`MemoryStreamManager`s main benefit is the reduced allocation. Object allocation is not a activity that should be taken lightly when high-performance is desired, especially because it's contra-part "GarbageCollection" is not cheap. What `MemoryStreamManager` does is to pool buffers of memory (on our case, 4096 bytes long) avoiding allocations as much as possible and reasonable. Whenever a larger buffer is required, `MemoryStreamManager` either combines small buffers or creates a `LargeBuffer`, which can also be pooled and are placed directly on GEN 2, which GC touches even less frequently. Proof of such optimised behaviour is the decrease activity by the GC (the yellow up-side-down "triangles" on the screen shot) and the reduced memory footprint of the game (489.5mb vs 504.9mb). And this far I've only touched the PNG class.

> The calls to `new MemoryStream()` in the `Save` method make sense as something that would benefit - but saving a PNG seems pretty rare - is there a scenario in the game you were trying to improve here?

I'm not trying to improve any specific scenario with the `Save` implementation, I was just "standardising" the whole class.

Leonardo-Ferreira committed:
License Header included

Leonardo-Ferreira commented:
@pchote 
> A couple of extra points, and I second RoosterDragon's questions - we should be only using MemoryStreams behind the load screen, where we are already generating significant amounts of garbage, which is cleaned up with a forced GC before the gameplay phase starts.

A couple of things:
1. MemoryStreams are being used all over the code, specially on the "EchoConnection" and "NetworkConnection". allocating those is not to be trifled with...
2. Forcing GC collections is not a recommended thing to do and per evidence it's not doing much (perhaps nothing at all), at least for my current setup (i5, 16gb ram). Out of curiosity, I disabled all forced GC collections and the memory profiling was 95% unaltered. I actually liked the memory footprint better with the forced collection off. 
3. Red Alert originally required 8mb of ram. It's ok to consume more and there's no point on restraining ourselves to the original requirement, but we're talking 30-40x the consumption for start-up alone!
4. I have to measure/understand how NetworkConnection behave for "server" scenario when there's several players on the game... Allocating MemoryStreams is not only something that consumes memory, it also consumes time, specially when you have memory fragmentation

> I would like to see some numbers quantifying the actual ingame benefit before we start any major efforts to rework the use of streams through the engine.

well, this doesn't have to be a "big-bang" implementation. We can do it piece-by-piece on a very calm and time-dilated manner. I don't see this as a "major effort rework" kind of thing since usage will change only slightly and the whole thing exposes the same api... apart from the "instantiation", nothing changes...  
As for the numbers, on the `PNG` alone we saw a 2-5% decrease on memory footprint and the GC activation dropped about 15%.

Leonardo-Ferreira commented:
@pchote, @RoosterDragon any other questions?

RoosterDragon commented:
Sure, we understand the premise that avoiding unnecessary allocations is good for performance. We would like to validate this PR is helping towards that, and measuring using only the memory graph is VS is a very blunt tool for knowing if those changes are due to this PR alone.

> The cleanest way to check the result is to actually dump the memory and perform a memory count but that would be VERY difficult because there are several memory streams all around and it be very hard to tell if that particular memory belong to a PNG or to another class...

There are tools for this. Since you mentioned VS, I suggest taking a look at:
https://docs.microsoft.com/en-us/visualstudio/profiling/memory-usage?view=vs-2019
https://docs.microsoft.com/en-us/visualstudio/profiling/dotnet-alloc-tool?view=vs-2019

The first will let you analyse objects on the heap. The second allows you to track all object allocations, and which methods allocated them.

The latter can be very helpful for focusing on the area you're trying to change and ignore the noise generated by everything else going on in the program. For example, if we run a similar allocation tool on the RA mod during startup only we can drill into the `Png` constructor to see it allocated 55Mb.

![image](https://user-images.githubusercontent.com/3399086/96751433-12cf6980-13c5-11eb-80e4-01153f187ba2.png)

If we run again on your PR, we see 55Mb allocated.

![image](https://user-images.githubusercontent.com/3399086/96750779-45c52d80-13c4-11eb-806b-d239e0aa7760.png)

This would seem to imply this PR isn't working as we hoped. We can see above that the memory streams aren't allocating much - instead it's the byte arrays the Png constructor creates directly.

Ok, so why aren't the memory streams actually allocating memory for us in the first place? Let's come back to my original comment:
> The calls to the new MemoryStream(byte[] buffer) overload uses the provided buffer as the backing array in the memory stream.

(as noted in [the docs](https://docs.microsoft.com/en-us/dotnet/api/system.io.memorystream.-ctor?view=netcore-3.1#System_IO_MemoryStream__ctor_System_Byte___))

Armed with this bit of knowledge, we can see why `MemoryStreamManager` has not helped us here. Its goal is to provide pooled buffers when a memory stream needs to allocate a new one internally. However, the memory streams we create during `Png` construction are provided with an external buffer to use instead - they don't need to allocate internal buffers.

Therefore, reducing allocations in the `Png` constructor shouldn't focus on memory streams, but instead on avoiding the `byte[]` allocations we need to make.

The PR is based on a sensible hypothesis (memory streams allocate memory) and a sensible change to tackle that hypothesis (use pooled buffers) - but by going through the validation steps above, we can see things weren't quite as we expected (the memory allocation was actually due to `byte[]` instead).

Hope that helps explain my initial query about the impact of this change.

Leonardo-Ferreira committed:
Implemented overload to get empty stream with desired length

Leonardo-Ferreira committed:
Refactor to allocation-free implementations

Leonardo-Ferreira committed:
Reduced allocation count of byte[]

Leonardo-Ferreira committed:
IHDR parsing is now a single allocation and allocation

Leonardo-Ferreira commented:
@RoosterDragon thank you for deep analysis. I must say im quite delighted with it. There are a few things I would like to point out though:

> This would seem to imply this PR isn't working as we hoped. We can see above that the memory streams aren't allocating much - instead it's the byte arrays the Png constructor creates directly.

Reduce memory usage is not the main advantage of `MemoryStreamManager`. Reducing the number of objects GC collects, thus reducing the total amount of time blocked by GC [IS THE MAIN](https://github.com/microsoft/Microsoft.IO.RecyclableMemoryStream#purpose) advantage. Specially of large arrays.

> [...] avoiding unnecessary allocations is good for performance. We would like to validate this PR is helping towards that [...]

Using the allocation tool I was able to determine that the `Byte[]` allocation dropped by 63 thousand allocations. `MemoryStream` dropped by almost 1k. 

GC objects collected dropped by 108k. Couldn't measure how much less time was spent on GC.

The "diagsession" files are over 300MB (comprise of starting up and quitting as soon as possible, 1minute 20 seconds of session) but if you want I can send them over, just give me an email so I can direct the "wetransfer" link...

I can't really explain why you're seeing the same 55mb though... I will try to reconcile it

Leonardo-Ferreira committed:
Expanded the MemoryStream impl to the PNG borders

Leonardo-Ferreira committed:
Map caching reads the filestream to memory all at once avoiding IO trips

Leonardo-Ferreira committed:
Delayed memoryStream instantiation

Leonardo-Ferreira committed:
Fine tuned the block sizes

Leonardo-Ferreira commented:
@RoosterDragon I've worked more on refinements and implemented some "allocation free" changes. I expanded the scope of the implementation a bit, taking it to the edges of the PNG class (and not only internally to the PNG class). I also implemented a minor change on how the PNG file is handled, loading it all at once to memory (instead of multiple IO trips)...

I fiddled a bit further and was able to reduce total allocations even further and the GC collection dropped even more, to a total of 162k less!

Would you mind revisiting this?  

I must confess that I develop from MACOSX using VS for Mac so I don't have such a great tooling in place (I asked a friend with VS Enterprise for the tests I mentioned above)

RoosterDragon commented:
I think it's clicking for me now, this buffer pooling (in the usage for this PR) doesn't actually get a chance to reuse buffers much (or we'd see reduced allocations overall) but at least the pooled buffers get promoted to Gen2, so Gen0 GCs have less crud to dig through? Got it.

My thoughts:
- Changing calls from use `new MemoryStream(someByteArray)` to `MemoryStreamManager.GetMemoryStream(someByteArray)`
Per the `RecyclableMemoryStream` docs:
`You can also provide an existing buffer. It’s important to note that this buffer will be copied into the pooled buffers owned by RecyclableMemoryStream`.
I don't understand the benefit of this. In our use cases we're usually just reading from that original buffer and then we're done with it and it can be reclaimed right away. We've already paid the cost of creating our own byte array, changing this call now just creates extra overhead of copying our first array to some pooled array - is there some benefit I'm missing here?
- That leaves the `Png` changes. Here you've now got `ms = PrepareMemoryStream(s, length);` so we can gain from the benefit of the pooling each time we use a new MemoryStream for reading parts of the `Png` file. This makes some sense to me, but seems like a halfway step - why not entirely ditch the `MemoryStream`s and just read directly from the source Stream? The only non-trivial bit comes in avoiding the `using (var ns = new MemoryStream(data.ToArray()))` call once the whole PNG is read. If you tracked the windows given in the `IDAT` chunks, you could seek back to them and use that to create a stream that only read the bytes within the `IDAT` sections, which you could then feed directly back into the `InflaterInputStream`.

If we did that, you get a `Png` reader capable of reading a source stream without having to use a `MemoryStream` at all. This means you avoid allocating the memory in the first place, which is even better for the GC, and we don't now have to worry about a global cache and all the downsides of cache management that entails.

Introducing a new dependency and a global cache via `RecyclableMemoryStream` is a fairly high bar of extra complexity to justify. Where local solutions exist (e.g. rewrite code to avoid the need for `MemoryStream`s in the first place) I'd prefer to do that first.

Leonardo-Ferreira commented:
>is there some benefit I'm missing here?

Our current architecture intends of the PNGs streams to live perpetually in-memory. The sooner they get out of `Gen 0` the better because it's less work for the GC to do. There were some instances where we were allocating `bytes[]` without actually having to, I spotted a few a changed those implementation for allocation free ones (marked with comments for future reference)

> This makes some sense to me, but seems like a halfway step - why not entirely ditch the MemoryStreams and just read directly from the source Stream?

Thats the plan, I was just afraid to do this huge implementation and being shut down because it's too huge of change (happened before in other projects), so I decided to take small steps, setup a common ground for understandment, and then go for the big fishes... if by any chance you fellows didn't approve of this change, I would at least have "lost" only a few hours of coding...

> The only non-trivial bit comes in avoiding the using (var ns = new MemoryStream(data.ToArray())) call once the whole PNG is read

Already worked around that with a allocation free implementation on commit 2a36c37

> [...] you get a Png reader capable of reading a source stream without having to use a MemoryStream at all. This means you avoid allocating the memory in the first place [...]

I don't think that's feasible. We could cut the `MemoryStream` but "under the hood" that means that either we would be working with the disk directly (IO trips all the time) or some other class that would abstract the handling of `byte[]`... come and thinking about it, the first option (working with the disk directly) is not really an option because data has to be in memory for processing...

>Introducing a new dependency and a global cache via RecyclableMemoryStream is a fairly high bar of extra complexity to justify

For this implementation alone, I agree, but remember, this is a first step, take a look at a bigger picture! RecyclableMemoryStreams can be used all over the code: networking, audio, crypt, replay recording and more!

pchote commented:
> Our current architecture intends of the PNGs streams to live perpetually in-memory. 

Actually not, and if something is causing this to happen in practice then we have a major regression! All textures are copied into a `Sheet` ASAP after loading (either directly for UI textures, or via a `SheetBuilder` atlas for gameplay artwork), which is then copied to the GPU and purged from RAM. Any references to the concrete file loaders like Png should be discarded almost immediately.

Leonardo-Ferreira commented:
> All textures are copied into a Sheet ASAP after loading

yes, I know, but in practice what happens is the "caching" of those bytes. Once read from disk/files they are only moved around (whether copied or "by ref") and they are not reclaimed. Moving them to the GPU is smart (very elegant actually, I haven't picked that out previously) but nonetheless they are still "loaded". As for the `PNG` instances if you trace the allocation and the destruction (via `~PNG()`) up until main screen displaying, you will find that there were only 13 instances automatically or forcibly finalised, and over 120 instances created... this may vary according to resource availability

> which is then copied to the GPU and purged from RAM

Not true for GPU-less environments? Or for integrated GPU also...

RoosterDragon commented:
For comparison, here's what I meant for a "local changes" example for PNG to avoid the need for a MemoryStream in the first place. https://github.com/OpenRA/OpenRA/compare/bleed...RoosterDragon:png-windows

I suggest that this is better in two aspects:
- This improves the GC impact by avoiding allocating the memory at all, rather than just trying to reduce the impact by having the memory living in Gen 2.
- We don't need any changes outside the Png class for this.

I feel this is a better balance than using `RecycableMemoryStream` to smooth over but not actually remove the underlying problem.

Leonardo-Ferreira commented:
Your `WindowedStream` class is a implementation for the idea of .Net Standard 2.1's [`Span<T>`](https://docs.microsoft.com/en-us/archive/msdn-magazine/2018/january/csharp-all-about-span-exploring-a-new-net-mainstay) and [`Memory<T>`](https://docs.microsoft.com/en-us/dotnet/api/system.memory-1?view=netcore-3.1).

I asked my friend for GC Collection data on your code, but I believe that your result will be worst than my proposition because by using a `Stream`, the allocations continues to happen in similar ways, and not getting promoted to GEN 2, they will trigger collection... take a look here:

current:  
`var line = new byte[srcStride];` (line 141)
previously:
`var line = ds.ReadBytes(srcStride);` (old line 144)

It feels like, as we Brazilian say, that you changed 6 for half-a-dozen or less... allocation is still happening, they're just being handled elsewhere... `ReadUInt8` is worse than `ReadByte` because, being a extension method, it reads a byte o some checking and then cast it back to byte! Most of the PNG also come from FileStreams,  and all those `ReadByte`/`ReadUInt8` will incur I/O, so the performance might be even worst than mine implementation (not worst than the original one)

I don't know, im really skeptic about this... since you provided no actual data to back your claim that this is in fact better I'll wait on my pal...

but hey, it's not like you need my approval to get this on bleed... this is only my 2 cents...

RoosterDragon commented:
> current:
> `var line = new byte[srcStride];` (line 141)
> previously:
> `var line = ds.ReadBytes(srcStride);` (old line 144)

Notice in the new version, this allocation is done outside the `for (var y = 0; y < Height; y++)` loop. This means you allocate `line` once for the whole image now, whereas previously it was allocated once per row.

----

`ReadUInt8` is slower in order to correctly check for the end-of-stream. None of the current `ReadByte` calls are checking for `-1` so would happily read garbage data instead of throwing an exception for the end of the stream instead. That is just a bug fix for a common mishap when using streams, and not intended to help performance.

Reading in larger batches per some of the changes in your PR could be worthwhile for streams that work better that way, and is still something to consider.

----

Some high-level comparisons, ra mod startup:

| Branch | Bytes Allocated via `Png` | Gen 0 collections |
|---|---:|---:|
| Bleed | 55 Mb | 105 |
| #18724 | 54 Mb | 100 |
| png-windows | 28 Mb | 98 |

pchote commented:
We removed a bunch of unnecessary buffer copies in the sheet loading code as part of the remastered asset prep, and its not clear from the discussion above exactly how we would want to proceed here. Closing to help clean up the PR queue.
